🎨  Improved Update Notifications

kirrg001 opened the pull request:
refs #5071 

We want to achieve these things:

1. Get rid of the hardcoded notification message - receive messages from the update check service.
2. Be able to receive custom messages.
3. Be less annoying and turn `dismissible` on by default.
4. Future proof dynamic (e.g. be able to receive multiple notifications, notification groups)
5. Custom Update Check endpoint for development.
6. We stick with top bar notification in LTS, because we only release LTS once per 1-3 month and with this PR we are able to send custom messages, but for 1.0 we will move default release notifications into the about page. And urgent/priority notifications will show up on the top bar e.g. security updates, migration notifications etc.

**Note**: I deleted the `displayUpdateNotification` settings key, because it's not used anymore. No need to add a migration still, as LTS ends anyway this year. 

**Note**: I've renamed `seenNotifications` to `notifications`. `notifications` is the true place to store/read notifications from. Unique identifier is a uuid. We simply store any new notification in it. If you remove a notification, it get's marked as seen. And deleted after 3 month automatically. No need to add a migration still, as LTS ends anyway this year - but we'll do for 1.x.

**Note**: No usage of notification in process store anymore, because this now that the messages are coming from the UNC, we can't simply store the messages in process. Because if you receive a notification and you restart Ghost, the message is lost. You would get it back on the next request to the service. No need to cache the notifications in general, because it's only one read operation to one key in the settings table.

**Note** Each commit message has all the content you need.

- [x] remove `config.updateCheck`, keep the privacy flag and remove this flag from Pro
- [x] add possibility to receive multiple release notifications (just in case we want that in the future)
- [x] fix tests / add tests
- [x] change privacy flag behaviour
- [x] care about restart use case (therefor we have to store notifications in the settings table)
- [x] add condition to receive custom pro notification for migrations only if blog is allowed to migrate -> generalise this by using e.g. config.allowedGroupMessage (an array of regex strings e.g. `migration` or `share-more`.
- [x] future case: be able to send a custom notification to a single blog on pro
- [x] code quality check/read commit messages again/overview of the last changes

ErisDS commented on core/server/data/schema/default-settings.json:
```diff
@@ -9,9 +9,6 @@
         },
-        "displayUpdateNotification": {
-            "defaultValue": null
-        },
```
How are we keeping track of notifications in this updated logic?

ErisDS commented:
I tried to add in a client extension for showing an extra notification from config. That got complicated quickly, and developing on LTS is so crazy hard now!

I would suggest that somewhere in this server side code where the notifications are fetched, it would be great to merge in `config.notifications` if it is set and an array. That would solve our one last case. I've tried to do this, but it ends up touching the same areas as you're already working on.

kirrg001 commented:
> would suggest that somewhere in this server side code where the notifications are fetched, it would be great to merge in config.notifications if it is set and is an array. That would solve our one last case. I've tried to do this, but it ends up touching the same areas as you're already working on.

Will check tomorrow.

kirrg001 committed:
Remove hardcoded notification in admin controller

- this is one of the most import changes
- we remove the hardcoded release message
- we also remove adding a notification manually in here, because this will work differently from now on
  -> you receive a notification (release or custom) in the update check module and this module adds the notification as is to our database

kirrg001 committed:
Change default core settings keys

- remove displayUpdateNotification
  -> this was used to store the release version number send from the UCS
  -> based on this value, Ghost creates a notification container with self defined values
  -> not needed anymore

- rename seenNotifications to notifications
  -> the new notifications key will hold both
     1. the notification from the USC
     2. the information about if a notification was seen or not

I personally think we don't have to cleanup these keys, as there are not used anymore.
If we want to cleanup, we have to add a migration. But as this is LTS and core settings are not imported, it doesn't matter i would say

kirrg001 committed:
Update Check Module: Request to the USC depends on the privacy configuration

- useUpdateCheck: true -> does a checkin in the USC (exposes data)
- useUpdateCheck: false -> does only a GET query to the USC (does not expose any data)
- make the request handling dynamic, so it depends on the flag
- add an extra logic to be able to define a custom USC endpoint (helpful for testing)

kirrg001 committed:
Update check module: re-work condition when a check should happen

- only if the env is not correct
- remove deprecated config.updateCheck
- remove isPrivacyDisabled check (handled differently now, explained in last commit)

kirrg001 committed:
Update check module: remove `showUpdateNotification` and readability

- showUpdateNotification was used in the admin controller to fetch the latest release version number from the db
- no need to check against semver in general, the USC takes care of that (no need to double check)
- improve readability of `nextUpdateCheck` condition

kirrg001 committed:
Update check module: refactor `updateCheckResponse`

- remove db call to displayUpdateNotification, not used anymore
- support receiving multiple custom notifications
- support custom notification groups
- the default group is `all` - this will always be consumed
- groups can be extended via config e.g. `notificationGroups: ['migration']`

kirrg001 committed:
Update check module: refactor createCustomNotification helper

- get rid of taking over notification duplication handling (this is not the task of the update check module)
- ensure we have good fallback values for non present attributes in a notification
- get rid of semver check (happens in the USC)

kirrg001 committed:
Refactor notification API

- reason: get rid of in process notification store
  -> this was an object hold in process
  -> everything get's lost after restart
  -> not helpful anymore, because imagine the following case
     -> you get a notification
     -> you store it in process
     -> you mark this notification as seen
     -> you restart Ghost, you will receive the same notification on the next check again
     -> because we are no longer have a separate seen notifications object
- use database settings key `notification` instead
- refactor all api endpoints to support reading and storing into the `notifications` object
- most important: notification deletion happens via a `seen` property (the notification get's physically deleted 3 month automatically)
  -> we have to remember a seen property, because otherwise you don't know which notification was already received/seen

kirrg001 committed:
Remove unused translation key

kirrg001 committed:
Add listener to remove seen notifications automatically after 3 month

- i just decided for 3 month (we can decrease?)
- at the end it doesn't really matter, as long as the windows is not tooooo short
- listen on updates for the notifications settings
- check if notification was seen and is older than 3 month

kirrg001 committed:
TESTS ✌🏻

kirrg001 commented:
Ready to review!
